blocks/extensions.py

```diff
@@ -0,0 +1,141 @@
+class TrainingExtension(object):
```

Comment:
My plan is the following: we can have a `Timer` extension that a writes message to the main log when a period of time has passed since the last timer activation. Other extensions can handle this message. One should be careful with the order in which extensions are called in such a scheme of course.